src/Language/Haskell/Liquid/Synthesize/Generate.hs

```diff
@@ -0,0 +1,266 @@
+    Nothing             -> return [] 
+    Just (resTy, subGs) -> 
+      do  let argCands = map (withSubgoal em0) subGs
+              toGen    = foldr (\x b -> (not . null) x && b) True argCands
+          es <- do  curExprId <- sExprId <$> get
+                    if toGen then 
+                      prune curExprId c argCands
+                      else return []
+          curExprId <- sExprId <$> get
+          let nextEm = map (resTy, , curExprId + 1) es
+          modify (\s -> s {sExprMem = nextEm ++ sExprMem s })
+          argsFill em0 cs (es ++ es0)
+
+withDepthFill :: SearchMode -> SpecType -> Int -> [(Type, GHC.CoreExpr, Int)] -> SM [CoreExpr]
+withDepthFill i t depth tmp = do
+  exprs <- fill i depth tmp []
+
+  filterElseM (hasType t) exprs $ 
+      if depth < maxAppDepth
+        then do modify (\s -> s { sExprId = sExprId s + 1 })
+                withDepthFill i t (depth + 1) tmp
+        else return []
+
+fill :: SearchMode -> Int -> [(Type, GHC.CoreExpr, Int)] -> [CoreExpr] -> SM [CoreExpr] 
+fill _ _     []                 accExprs 
+  = return accExprs
+fill i depth (c : cs) accExprs 
+  = case subgoals (fst3 c) of 
+      Nothing             -> return [] -- Not a function type
+      Just (resTy, subGs) ->
+        do  specSubGs <- liftCG $ mapM trueTy (filter (not . isFunction) subGs)
+            mapM_ genArgs specSubGs
+            em <- sExprMem <$> get
+            let argCands  = map (withSubgoal em) subGs
+                toGen    = foldr (\x b -> (not . null) x && b) True argCands
+            newExprs <- do  curExprId <- sExprId <$> get 
+                            if toGen then 
+                              prune curExprId c argCands
+                              else return []
+            curExprId <- sExprId <$> get
+            let nextEm = map (resTy, , curExprId + 1) newExprs
+            modify (\s -> s {sExprMem = nextEm ++ sExprMem s }) 
+            let accExprs' = newExprs ++ accExprs
+            fill i depth cs accExprs' 
+
+-------------------------------------------------------------------------------------------
+-- |                       Pruning terms for function application                      | --
+-------------------------------------------------------------------------------------------
+type Depth = Int
+
+feasible :: Depth -> (CoreExpr, Int) -> Bool
+feasible d c = snd c >= d
+
+feasibles :: Depth -> Int -> [(CoreExpr, Int)] -> [Int]
+feasibles _ _ []
+  = []
+feasibles d i (c:cs) 
+  = if feasible d c 
+      then i : feasibles d (i+1) cs
+      else feasibles d (i+1) cs
+
+isFeasible :: Depth -> [[(CoreExpr, Int)]] -> [[Int]]
+isFeasible d =  map (feasibles d 0)
+
+toIxs :: Int -> [[Int]] -> [Int]
+toIxs _ [] 
+  = [] 
+toIxs i (f:fs)
+  = if null f 
+      then toIxs (i+1) fs
+      else i : toIxs (i+1) fs
+
+findFeasibles :: Depth -> [[(CoreExpr, Int)]] -> ([[Int]], [Int])
+findFeasibles d cs = (fs, ixs)
+  where fs  = isFeasible d cs
+        ixs = toIxs 0 fs
+
```

Comment:
@pbougou the above should be something like `ixs = [i | (i,f)<- zip [0..] fs, not (null f)]`